README.md
# Antlr4 CodeGenerator tool

[![License](https://img.shields.io/badge/license-MIT-blue.svg?logo=data%3Aimage%2Fpng%3Bbase64%2CiVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAAHYcAAB2HAY%2Fl8WUAAAAZdEVYdFNvZnR3YXJlAHBhaW50Lm5ldCA0LjAuMTCtCgrAAAADB0lEQVR4XtWagXETMRREUwIlUAIlUAodQAl0AJ1AB9BB6AA6gA6MduKbkX%2BevKecNk525jHO3l%2Fp686xlJC70%2Bl0C942vjV%2Bn9FreVQbBc0wWujfRpW8Z78JaIb53hhJ1ygTA80w9PQ36duBMjHQHPCuoQZfutSjeqU1PAJN4E3j2pN7aVKv6pnWcgGawNfGa5N6prVcgGZBn8yvVXZXQbOgPXokXaPMNZwoc41D%2FaHZ8b7hpBrKjnCizIjD%2FaHZ8aPR6%2BeZXqqh7Agnyow43B%2BaZz40qnQ36a6rlsYgnChDLOkPzTN1z%2B9PafU0N3OAcaIMsaQ%2FNBufG1X9JyrtDMr0Y4xwokxlWX%2BPjAYdemhPrWeDvYcPJ8r0LO3v4oszNfivQQuTp2u9qJGKE2V6lvZ38UVj9q3t3oqEE2U2lvfXF4t6qPjTqDUV1fRyhw8nymws768vfOr2NtqOqFY4UUZE%2BusL6VDRX7%2FGzOHDiTIi0t9WMPsUKzNPx4kysf62gmuHir3sPXw4USbWny485ZOc2PsJ7VTro%2F3pwp5DxV7qHq2xa41TrY%2F2J7PfJkaHir3UwwdtU061PtqfTP0CUaYm2v3LxCtoDI2lMWk8p1of7Y8K0jhRJgaaYZwoE0P%2FpFUndZqtP6T4BE2zC5qtP6T4BE2zC5qtPyRN8OvhZUQae3ZBtT7anyb49PA6Ivp5wKnWR%2FvbJkncZXr6wokysf62CXRCWjmJxhqd2JwoE%2BuvTqS37JGJlB39GLzhRJmN5f31gz8XTpSJgWYYJ8rEQDOME2VioBnGiTIx0AzjRJkYaIZxokwMNMM4USYGmmGcKBMDzTBOlImBZhgnysRAM4wTZWKgGcaJMjHQDONEmRhohnGiTAw0wzhRJgaaYZwoEwPNME6UiYFmGCfKxEAzjBNlYqAZxokyMdAMoL%2FO%2BNi4bzjpT1e%2BNFb8V7gFzUXMLHqk%2BM1A8wArFj1S5GagOUly0SMtuxloTnJrUU%2B7QXOSW4t62g2ak9xa1NNu0Jzk1qKednK6%2Bw9roIB8keT%2F3QAAAABJRU5ErkJggg%3D%3D)](LICENSE.md)

[![Latest Release](https://img.shields.io/nuget/v/Antlr4CodeGenerator.Tool?logo=nuget&label=release)](https://www.nuget.org/packages/Antlr4CodeGenerator.Tool)
[![Latest pre Release](https://img.shields.io/nuget/vpre/Antlr4CodeGenerator.Tool?logo=nuget)](https://www.nuget.org/packages/Antlr4CodeGenerator.Tool)

[![Build Status master](https://dev.azure.com/pbruch0060/pbruch/_apis/build/status/ptr1120.Antlr4.CodeGenerator.Tool?branchName=master)](https://dev.azure.com/pbruch0060/pbruch/_build/latest?definitionId=1&branchName=master)
[![Build Status develop](https://dev.azure.com/pbruch0060/pbruch/_apis/build/status/ptr1120.Antlr4.CodeGenerator.Tool?branchName=develop)](https://dev.azure.com/pbruch0060/pbruch/_build/latest?definitionId=1&branchName=develop)



Just a commandline wrapper around the ANTLR (Java) tool ([Version 4.13.1](https://www.antlr.org/download/antlr-4.13.1-complete.jar)) for generating grammar artifacts.

The ANTLR (Java) tool is *bundled*, so no need to download it.

## Dependencies

* Java runtime

## Installation

* Install as dotnet **global tool**:

  ```shell
  dotnet tool install --global Antlr4CodeGenerator.Tool
  ```

* Install as dotnet **local tool** (inside a dotnet project where you want to generate language artifacts):
    * See also [sample project](samples/Calculator)
    * Enable local dotnet tools for your project by the following command on solution file directory level:

    ```shell
    dotnet new tool-manifest
    ```

    * Install as local tool to your project:

    ```shell
    dotnet tool install Antlr4CodeGenerator.Tool
    ```

## Usage

This tool only proxies the arguments to the ANTLR (Java) tool. (See also [ANTLR (Java) tool arguments documentation](https://github.com/antlr/antlr4/blob/master/doc/tool-options.md))

* **global tool**:

    ```shell
    dotnet antlr4-tool -Dlanguage=CSharp MyGrammar.g4
    ```

* **local tool** via MsBuild target:
    * See also [sample project](samples/Calculator)

    ```xml
    <Target Name="GenerateAntlrArtifacts" BeforeTargets="BeforeResolveReferences">
      <PropertyGroup>
        <_GrammarFile>$(ProjectDir)calculator.g4</_GrammarFile>
        <_Generated>$(ProjectDir).generated\</_Generated>
        <_Namespace>MyCalculator</_Namespace>
      </PropertyGroup>
       <Exec Command="dotnet antlr4-tool -Dlanguage=CSharp  -o &quot;$(_Generated)&quot; -visitor -listener -package $(_Namespace) &quot;$(_GrammarFile)&quot;" />
    </Target>
    ```


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [vNext]
- Updated dependencies:
    - Antlr 4.12.0 -> 4.13.1
## [2.1.0] / 2023-04-01
- Updated dependencies:
    - Antlr 4.11.1 -> 4.12.0
## [2.0.0] / 2022-10-29
- Updated dependencies:
  - netcore3.1 -> net6,
  - Antlr 4.8 -> 4.11.1.
- Improved documentation for handling possible spaces in tool directory arguments
## [1.2.0] / 2020-05-09
- Improved ci support.
- Added symbols nuget.
## [1.1.0] / 2020-05-08
- Renamed PackageId: Antlr4.CodeGenerator.Tool to Antlr4CodeGenerator.Tool because Antlr4.CodeGenerator package ID has been reserved (reserved namespace from the owner of the reserved prefix).
## [1.0.0] / 2020-05-08
- Initial release

[vNext]: https://github.com/ptr1120/Antlr4.CodeGenerator.Tool/compare/2.0.0...HEAD
[2.0.0]: https://github.com/ptr1120/Antlr4.CodeGenerator.Tool/compare/1.2.0...2.0.0
[1.2.0]: https://github.com/ptr1120/Antlr4.CodeGenerator.Tool/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/ptr1120/Antlr4.CodeGenerator.Tool/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/ptr1120/Antlr4.CodeGenerator.Tool/tree/1.0.0



samples/Calculator/generated/calculatorVisitor.cs
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     ANTLR Version: 4.13.1
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

// Generated from calculator.g4 by ANTLR 4.13.1

// Unreachable code detected
#pragma warning disable 0162
// The variable '...' is assigned but its value is never used
#pragma warning disable 0219
// Missing XML comment for publicly visible type or member '...'
#pragma warning disable 1591
// Ambiguous reference in cref attribute
#pragma warning disable 419

namespace MyCalculator {
using Antlr4.Runtime.Misc;
using Antlr4.Runtime.Tree;
using IToken = Antlr4.Runtime.IToken;

/// <summary>
/// This interface defines a complete generic visitor for a parse tree produced
/// by <see cref="calculatorParser"/>.
/// </summary>
/// <typeparam name="Result">The return type of the visit operation.</typeparam>
[System.CodeDom.Compiler.GeneratedCode("ANTLR", "4.13.1")]
[System.CLSCompliant(false)]
public interface IcalculatorVisitor<Result> : IParseTreeVisitor<Result> {
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.equation"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitEquation([NotNull] calculatorParser.EquationContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.expression"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitExpression([NotNull] calculatorParser.ExpressionContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.multiplyingExpression"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitMultiplyingExpression([NotNull] calculatorParser.MultiplyingExpressionContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.powExpression"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitPowExpression([NotNull] calculatorParser.PowExpressionContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.signedAtom"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitSignedAtom([NotNull] calculatorParser.SignedAtomContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.atom"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitAtom([NotNull] calculatorParser.AtomContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.scientific"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitScientific([NotNull] calculatorParser.ScientificContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.constant"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitConstant([NotNull] calculatorParser.ConstantContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.variable"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitVariable([NotNull] calculatorParser.VariableContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.func"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitFunc([NotNull] calculatorParser.FuncContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.funcname"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitFuncname([NotNull] calculatorParser.FuncnameContext context);
	/// <summary>
	/// Visit a parse tree produced by <see cref="calculatorParser.relop"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	/// <return>The visitor result.</return>
	Result VisitRelop([NotNull] calculatorParser.RelopContext context);
}
} // namespace MyCalculator


samples/Calculator/generated/calculatorListener.cs
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     ANTLR Version: 4.13.1
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

// Generated from calculator.g4 by ANTLR 4.13.1

// Unreachable code detected
#pragma warning disable 0162
// The variable '...' is assigned but its value is never used
#pragma warning disable 0219
// Missing XML comment for publicly visible type or member '...'
#pragma warning disable 1591
// Ambiguous reference in cref attribute
#pragma warning disable 419

namespace MyCalculator {
using Antlr4.Runtime.Misc;
using IParseTreeListener = Antlr4.Runtime.Tree.IParseTreeListener;
using IToken = Antlr4.Runtime.IToken;

/// <summary>
/// This interface defines a complete listener for a parse tree produced by
/// <see cref="calculatorParser"/>.
/// </summary>
[System.CodeDom.Compiler.GeneratedCode("ANTLR", "4.13.1")]
[System.CLSCompliant(false)]
public interface IcalculatorListener : IParseTreeListener {
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.equation"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterEquation([NotNull] calculatorParser.EquationContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.equation"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitEquation([NotNull] calculatorParser.EquationContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.expression"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterExpression([NotNull] calculatorParser.ExpressionContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.expression"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitExpression([NotNull] calculatorParser.ExpressionContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.multiplyingExpression"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterMultiplyingExpression([NotNull] calculatorParser.MultiplyingExpressionContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.multiplyingExpression"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitMultiplyingExpression([NotNull] calculatorParser.MultiplyingExpressionContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.powExpression"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterPowExpression([NotNull] calculatorParser.PowExpressionContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.powExpression"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitPowExpression([NotNull] calculatorParser.PowExpressionContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.signedAtom"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterSignedAtom([NotNull] calculatorParser.SignedAtomContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.signedAtom"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitSignedAtom([NotNull] calculatorParser.SignedAtomContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.atom"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterAtom([NotNull] calculatorParser.AtomContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.atom"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitAtom([NotNull] calculatorParser.AtomContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.scientific"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterScientific([NotNull] calculatorParser.ScientificContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.scientific"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitScientific([NotNull] calculatorParser.ScientificContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.constant"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterConstant([NotNull] calculatorParser.ConstantContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.constant"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitConstant([NotNull] calculatorParser.ConstantContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.variable"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterVariable([NotNull] calculatorParser.VariableContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.variable"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitVariable([NotNull] calculatorParser.VariableContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.func"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterFunc([NotNull] calculatorParser.FuncContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.func"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitFunc([NotNull] calculatorParser.FuncContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.funcname"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterFuncname([NotNull] calculatorParser.FuncnameContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.funcname"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitFuncname([NotNull] calculatorParser.FuncnameContext context);
	/// <summary>
	/// Enter a parse tree produced by <see cref="calculatorParser.relop"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void EnterRelop([NotNull] calculatorParser.RelopContext context);
	/// <summary>
	/// Exit a parse tree produced by <see cref="calculatorParser.relop"/>.
	/// </summary>
	/// <param name="context">The parse tree.</param>
	void ExitRelop([NotNull] calculatorParser.RelopContext context);
}
} // namespace MyCalculator


samples/Calculator/generated/calculatorLexer.cs
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     ANTLR Version: 4.13.1
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

// Generated from calculator.g4 by ANTLR 4.13.1

// Unreachable code detected
#pragma warning disable 0162
// The variable '...' is assigned but its value is never used
#pragma warning disable 0219
// Missing XML comment for publicly visible type or member '...'
#pragma warning disable 1591
// Ambiguous reference in cref attribute
#pragma warning disable 419

namespace MyCalculator {
using System;
using System.IO;
using System.Text;
using Antlr4.Runtime;
using Antlr4.Runtime.Atn;
using Antlr4.Runtime.Misc;
using DFA = Antlr4.Runtime.Dfa.DFA;

[System.CodeDom.Compiler.GeneratedCode("ANTLR", "4.13.1")]
[System.CLSCompliant(false)]
public partial class calculatorLexer : Lexer {
	protected static DFA[] decisionToDFA;
	protected static PredictionContextCache sharedContextCache = new PredictionContextCache();
	public const int
		COS=1, SIN=2, TAN=3, ACOS=4, ASIN=5, ATAN=6, LN=7, LOG=8, SQRT=9, LPAREN=10, 
		RPAREN=11, PLUS=12, MINUS=13, TIMES=14, DIV=15, GT=16, LT=17, EQ=18, COMMA=19, 
		POINT=20, POW=21, PI=22, EULER=23, I=24, VARIABLE=25, SCIENTIFIC_NUMBER=26, 
		WS=27;
	public static string[] channelNames = {
		"DEFAULT_TOKEN_CHANNEL", "HIDDEN"
	};

	public static string[] modeNames = {
		"DEFAULT_MODE"
	};

	public static readonly string[] ruleNames = {
		"COS", "SIN", "TAN", "ACOS", "ASIN", "ATAN", "LN", "LOG", "SQRT", "LPAREN", 
		"RPAREN", "PLUS", "MINUS", "TIMES", "DIV", "GT", "LT", "EQ", "COMMA", 
		"POINT", "POW", "PI", "EULER", "I", "VARIABLE", "VALID_ID_START", "VALID_ID_CHAR", 
		"SCIENTIFIC_NUMBER", "NUMBER", "E1", "E2", "SIGN", "WS"
	};


	public calculatorLexer(ICharStream input)
	: this(input, Console.Out, Console.Error) { }

	public calculatorLexer(ICharStream input, TextWriter output, TextWriter errorOutput)
	: base(input, output, errorOutput)
	{
		Interpreter = new LexerATNSimulator(this, _ATN, decisionToDFA, sharedContextCache);
	}

	private static readonly string[] _LiteralNames = {
		null, "'cos'", "'sin'", "'tan'", "'acos'", "'asin'", "'atan'", "'ln'", 
		"'log'", "'sqrt'", "'('", "')'", "'+'", "'-'", "'*'", "'/'", "'>'", "'<'", 
		"'='", "','", "'.'", "'^'", "'pi'", null, "'i'"
	};
	private static readonly string[] _SymbolicNames = {
		null, "COS", "SIN", "TAN", "ACOS", "ASIN", "ATAN", "LN", "LOG", "SQRT", 
		"LPAREN", "RPAREN", "PLUS", "MINUS", "TIMES", "DIV", "GT", "LT", "EQ", 
		"COMMA", "POINT", "POW", "PI", "EULER", "I", "VARIABLE", "SCIENTIFIC_NUMBER", 
		"WS"
	};
	public static readonly IVocabulary DefaultVocabulary = new Vocabulary(_LiteralNames, _SymbolicNames);

	[NotNull]
	public override IVocabulary Vocabulary
	{
		get
		{
			return DefaultVocabulary;
		}
	}

	public override string GrammarFileName { get { return "calculator.g4"; } }

	public override string[] RuleNames { get { return ruleNames; } }

	public override string[] ChannelNames { get { return channelNames; } }

	public override string[] ModeNames { get { return modeNames; } }

	public override int[] SerializedAtn { get { return _serializedATN; } }

	static calculatorLexer() {
		decisionToDFA = new DFA[_ATN.NumberOfDecisions];
		for (int i = 0; i < _ATN.NumberOfDecisions; i++) {
			decisionToDFA[i] = new DFA(_ATN.GetDecisionState(i), i);
		}
	}
	private static int[] _serializedATN = {
		4,0,27,189,6,-1,2,0,7,0,2,1,7,1,2,2,7,2,2,3,7,3,2,4,7,4,2,5,7,5,2,6,7,
		6,2,7,7,7,2,8,7,8,2,9,7,9,2,10,7,10,2,11,7,11,2,12,7,12,2,13,7,13,2,14,
		7,14,2,15,7,15,2,16,7,16,2,17,7,17,2,18,7,18,2,19,7,19,2,20,7,20,2,21,
		7,21,2,22,7,22,2,23,7,23,2,24,7,24,2,25,7,25,2,26,7,26,2,27,7,27,2,28,
		7,28,2,29,7,29,2,30,7,30,2,31,7,31,2,32,7,32,1,0,1,0,1,0,1,0,1,1,1,1,1,
		1,1,1,1,2,1,2,1,2,1,2,1,3,1,3,1,3,1,3,1,3,1,4,1,4,1,4,1,4,1,4,1,5,1,5,
		1,5,1,5,1,5,1,6,1,6,1,6,1,7,1,7,1,7,1,7,1,8,1,8,1,8,1,8,1,8,1,9,1,9,1,
		10,1,10,1,11,1,11,1,12,1,12,1,13,1,13,1,14,1,14,1,15,1,15,1,16,1,16,1,
		17,1,17,1,18,1,18,1,19,1,19,1,20,1,20,1,21,1,21,1,21,1,22,1,22,1,23,1,
		23,1,24,1,24,5,24,140,8,24,10,24,12,24,143,9,24,1,25,3,25,146,8,25,1,26,
		1,26,3,26,150,8,26,1,27,1,27,1,27,3,27,155,8,27,1,27,3,27,158,8,27,1,27,
		1,27,3,27,162,8,27,1,28,4,28,165,8,28,11,28,12,28,166,1,28,1,28,4,28,171,
		8,28,11,28,12,28,172,3,28,175,8,28,1,29,1,29,1,30,1,30,1,31,1,31,1,32,
		4,32,184,8,32,11,32,12,32,185,1,32,1,32,0,0,33,1,1,3,2,5,3,7,4,9,5,11,
		6,13,7,15,8,17,9,19,10,21,11,23,12,25,13,27,14,29,15,31,16,33,17,35,18,
		37,19,39,20,41,21,43,22,45,23,47,24,49,25,51,0,53,0,55,26,57,0,59,0,61,
		0,63,0,65,27,1,0,3,3,0,65,90,95,95,97,122,2,0,43,43,45,45,3,0,9,10,13,
		13,32,32,191,0,1,1,0,0,0,0,3,1,0,0,0,0,5,1,0,0,0,0,7,1,0,0,0,0,9,1,0,0,
		0,0,11,1,0,0,0,0,13,1,0,0,0,0,15,1,0,0,0,0,17,1,0,0,0,0,19,1,0,0,0,0,21,
		1,0,0,0,0,23,1,0,0,0,0,25,1,0,0,0,0,27,1,0,0,0,0,29,1,0,0,0,0,31,1,0,0,
		0,0,33,1,0,0,0,0,35,1,0,0,0,0,37,1,0,0,0,0,39,1,0,0,0,0,41,1,0,0,0,0,43,
		1,0,0,0,0,45,1,0,0,0,0,47,1,0,0,0,0,49,1,0,0,0,0,55,1,0,0,0,0,65,1,0,0,
		0,1,67,1,0,0,0,3,71,1,0,0,0,5,75,1,0,0,0,7,79,1,0,0,0,9,84,1,0,0,0,11,
		89,1,0,0,0,13,94,1,0,0,0,15,97,1,0,0,0,17,101,1,0,0,0,19,106,1,0,0,0,21,
		108,1,0,0,0,23,110,1,0,0,0,25,112,1,0,0,0,27,114,1,0,0,0,29,116,1,0,0,
		0,31,118,1,0,0,0,33,120,1,0,0,0,35,122,1,0,0,0,37,124,1,0,0,0,39,126,1,
		0,0,0,41,128,1,0,0,0,43,130,1,0,0,0,45,133,1,0,0,0,47,135,1,0,0,0,49,137,
		1,0,0,0,51,145,1,0,0,0,53,149,1,0,0,0,55,151,1,0,0,0,57,164,1,0,0,0,59,
		176,1,0,0,0,61,178,1,0,0,0,63,180,1,0,0,0,65,183,1,0,0,0,67,68,5,99,0,
		0,68,69,5,111,0,0,69,70,5,115,0,0,70,2,1,0,0,0,71,72,5,115,0,0,72,73,5,
		105,0,0,73,74,5,110,0,0,74,4,1,0,0,0,75,76,5,116,0,0,76,77,5,97,0,0,77,
		78,5,110,0,0,78,6,1,0,0,0,79,80,5,97,0,0,80,81,5,99,0,0,81,82,5,111,0,
		0,82,83,5,115,0,0,83,8,1,0,0,0,84,85,5,97,0,0,85,86,5,115,0,0,86,87,5,
		105,0,0,87,88,5,110,0,0,88,10,1,0,0,0,89,90,5,97,0,0,90,91,5,116,0,0,91,
		92,5,97,0,0,92,93,5,110,0,0,93,12,1,0,0,0,94,95,5,108,0,0,95,96,5,110,
		0,0,96,14,1,0,0,0,97,98,5,108,0,0,98,99,5,111,0,0,99,100,5,103,0,0,100,
		16,1,0,0,0,101,102,5,115,0,0,102,103,5,113,0,0,103,104,5,114,0,0,104,105,
		5,116,0,0,105,18,1,0,0,0,106,107,5,40,0,0,107,20,1,0,0,0,108,109,5,41,
		0,0,109,22,1,0,0,0,110,111,5,43,0,0,111,24,1,0,0,0,112,113,5,45,0,0,113,
		26,1,0,0,0,114,115,5,42,0,0,115,28,1,0,0,0,116,117,5,47,0,0,117,30,1,0,
		0,0,118,119,5,62,0,0,119,32,1,0,0,0,120,121,5,60,0,0,121,34,1,0,0,0,122,
		123,5,61,0,0,123,36,1,0,0,0,124,125,5,44,0,0,125,38,1,0,0,0,126,127,5,
		46,0,0,127,40,1,0,0,0,128,129,5,94,0,0,129,42,1,0,0,0,130,131,5,112,0,
		0,131,132,5,105,0,0,132,44,1,0,0,0,133,134,3,61,30,0,134,46,1,0,0,0,135,
		136,5,105,0,0,136,48,1,0,0,0,137,141,3,51,25,0,138,140,3,53,26,0,139,138,
		1,0,0,0,140,143,1,0,0,0,141,139,1,0,0,0,141,142,1,0,0,0,142,50,1,0,0,0,
		143,141,1,0,0,0,144,146,7,0,0,0,145,144,1,0,0,0,146,52,1,0,0,0,147,150,
		3,51,25,0,148,150,2,48,57,0,149,147,1,0,0,0,149,148,1,0,0,0,150,54,1,0,
		0,0,151,161,3,57,28,0,152,155,3,59,29,0,153,155,3,61,30,0,154,152,1,0,
		0,0,154,153,1,0,0,0,155,157,1,0,0,0,156,158,3,63,31,0,157,156,1,0,0,0,
		157,158,1,0,0,0,158,159,1,0,0,0,159,160,3,57,28,0,160,162,1,0,0,0,161,
		154,1,0,0,0,161,162,1,0,0,0,162,56,1,0,0,0,163,165,2,48,57,0,164,163,1,
		0,0,0,165,166,1,0,0,0,166,164,1,0,0,0,166,167,1,0,0,0,167,174,1,0,0,0,
		168,170,5,46,0,0,169,171,2,48,57,0,170,169,1,0,0,0,171,172,1,0,0,0,172,
		170,1,0,0,0,172,173,1,0,0,0,173,175,1,0,0,0,174,168,1,0,0,0,174,175,1,
		0,0,0,175,58,1,0,0,0,176,177,5,69,0,0,177,60,1,0,0,0,178,179,5,101,0,0,
		179,62,1,0,0,0,180,181,7,1,0,0,181,64,1,0,0,0,182,184,7,2,0,0,183,182,
		1,0,0,0,184,185,1,0,0,0,185,183,1,0,0,0,185,186,1,0,0,0,186,187,1,0,0,
		0,187,188,6,32,0,0,188,66,1,0,0,0,11,0,141,145,149,154,157,161,166,172,
		174,185,1,6,0,0
	};

	public static readonly ATN _ATN =
		new ATNDeserializer().Deserialize(_serializedATN);


}
} // namespace MyCalculator
